File: Leetcode/src/main/java/candy/Candy.java
package candy;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;

public class Candy {
    public int candy(@NotNull int[] ratings) {
        int len = ratings.length;
        int[] candiesArray = new int[len];
        Arrays.fill(candiesArray, 1);
        for (int index = 1; index < len; index++) {
            if (ratings[index] > ratings[index - 1]) {
                candiesArray[index] = candiesArray[index - 1] + 1;
            }
        }
        for (int index = len - 2; index >= 0; index--) {
            if (ratings[index] > ratings[index + 1]) {
                candiesArray[index] = Math.max(candiesArray[index], candiesArray[index + 1] + 1);
            }
        }
        int totalCandies = 0;
        for (int candies : candiesArray) {
            totalCandies += candies;
        }
        return totalCandies;
    }
}

File: Leetcode/src/main/kotlin/candy/Candy.kt
package candy

class Candy {
    fun candy(ratings: IntArray): Int {
        val len = ratings.size
        val candiesArray = MutableList(len) { 1 }
        for (index in 1 until len) {
            if (ratings[index] > ratings[index - 1]) {
                candiesArray[index] = candiesArray[index - 1] + 1
            }
        }
        for (index in len - 2 downTo 0) {
            if (ratings[index] > ratings[index + 1]) {
                candiesArray[index] = candiesArray[index].coerceAtLeast(candiesArray[index + 1] + 1)
            }
        }
        return candiesArray.reduce { totalCandies, candies ->
            totalCandies + candies
        }
    }
}
